File: 2-add-two-numbers/solution.py
from typing import Optional


class ListNode:
  def __init__(self, val=0, next=Optional['ListNode']):
    self.val = val
    self.next = next


class Solution:
  @staticmethod
  def one_step(l1: Optional[ListNode], l2: Optional[ListNode], leftover=0) -> tuple[ListNode, int]:
    val1, val2 = 0 if l1 is None else l1.val, 0 if l2 is None else l2.val
    val = val1 + val2 + leftover
    return ListNode(val % 10 if val >= 10 else val), 1 if val >= 10 else 0

  def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
    curr1, curr2 = l1, l2
    result, prev = None, None
    leftover = 0
    while curr1 is not None or curr2 is not None or leftover:
      new_node, leftover = Solution.one_step(curr1, curr2, leftover)
      curr1, curr2 = None if curr1 is None else curr1.next, None if curr2 is None else curr2.next

      if result is None:
        result = prev = new_node
        continue

      prev.next = new_node
      prev = new_node

    return result

File: 2-add-two-numbers/solution.cpp
#include <tuple>

struct ListNode
{
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
private:
  static std::tuple<ListNode *, int> one_step(const ListNode *l1, const ListNode *l2, const int leftover = 0)
  {
    const int val1 = l1 == nullptr ? 0 : l1->val, val2 = l2 == nullptr ? 0 : l2->val;
    const int val = val1 + val2 + leftover;
    return {new ListNode(val >= 10 ? val % 10 : val), val >= 10 ? 1 : 0};
  }

public:
  ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
  {
    const auto *curr1 = l1;
    const auto *curr2 = l2;
    ListNode *result = nullptr;
    ListNode *prev = nullptr;
    int leftover = 0;
    while (curr1 != nullptr || curr2 != nullptr || leftover)
    {
      const auto step = Solution::one_step(curr1, curr2, leftover);
      auto *new_node = std::get<0>(step);
      leftover = std::get<1>(step);

      curr1 = curr1 == nullptr ? nullptr : curr1->next;
      curr2 = curr2 == nullptr ? nullptr : curr2->next;

      if (result == nullptr)
      {
        result = prev = new_node;
        continue;
      }

      prev->next = new_node;
      prev = new_node;
    }

    return result;
  }
};